### DISCUSSION TITLE
Get bluetooth controllers link key

### DISCUSSION THREAD

**Discussion by grimpy** [ndeadly/MissionControl#499] [Q&A] [answered]
I have read that getting exporting and importing the controller link keys is on the roadmap. 

Could you provide information how to retreive this information manually?
I would like to take the link key from the switch and use the same one on my pc.

**Comment by ndeadly** [comment, marked as answer]
The database entry containing the link key can be retrieved via homebrew by calling either [`btdrvGetPairedDeviceInfo`](https://github.com/switchbrew/libnx/blob/8b17648d0b910aa8496efcc7a5c8f40d5d15e9db/nx/include/switch/services/btdrv.h#L649) for an individual controller, or [`setsysGetBluetoothDevicesSettings`](https://github.com/switchbrew/libnx/blob/8b17648d0b910aa8496efcc7a5c8f40d5d15e9db/nx/include/switch/services/set.h#L1023) to retrieve the entire database. In both cases you will receive [`SetSysBluetoothDevicesSettings`](https://github.com/switchbrew/libnx/blob/8b17648d0b910aa8496efcc7a5c8f40d5d15e9db/nx/include/switch/services/set.h#L313-L339) structures that contain the link key. It is also possible to extract this info without writing code by dumping the system save that contains this data and locating the key at the correct location in a hex editor, but I will leave this an an exercise for the reader as I don't know the specifics.

If you plan to synchronise with your PC, you will also need to change/spoof its Bluetooth host address to that of your Switch in order for the controller to match the link key to the correct device. This can be retrieve by calling [`btdrvGetAdapterProperty`](https://github.com/switchbrew/libnx/blob/8b17648d0b910aa8496efcc7a5c8f40d5d15e9db/nx/include/switch/services/btdrv.h#L479) with [`BtdrvAdapterPropertyType_Address`](https://github.com/switchbrew/libnx/blob/8b17648d0b910aa8496efcc7a5c8f40d5d15e9db/nx/include/switch/services/btdrv_types.h#L21) as the type argument. Or, if you're lazy, my [btdb ](https://github.com/ndeadly/MissionControl/files/7227157/btdb-1300.zip)application will display it to the screen. Alternatively, you can spoof your Switch's host address to that of your PC using mission control's config .ini file. Note, you will have to reboot and re-pair the controller to get it to store the updated host address before grabbing the link key info.

**Reply by grimpy** [reply]
Thanks for the reply. I've never compiled stuff for the switch before so I'll take some time to set that up and give it a trial. My initial thought was to quickly hack your btdb tool to print out this extra information, but I can't seem to find the sourcecode of it in this repo. Is the code available somewhere?

**Reply by ndeadly** [reply]
btdb started off as a bit of an exploratory exercise before the bluetooth services were well documented, and was quickly hacked to display the database info to aid with controller identification. It was never really intended for any kind of release, so the code remains an unstructured mess of commented code and other unnecessary stuff that I never bothered to clean up. This is the only reason you won't find the source online anywhere. 

It should be pretty straightforward to modify the switchbrew [hello world](https://github.com/switchbrew/switch-examples/tree/master/graphics/printing/hello-world) example to fetch and print the link keys, but I can also clean up the btdb code and make a public gist out of it later today if you want.

**Reply by grimpy** [reply]
i already got a very rudimentary version working (https://gist.github.com/grimpy/dba849e7583a6c24b30c736f167156af) very very drafty but does the job for now.

I just noticed my lovely bt dongle on my PC is refusing to get a new mac address assigned so Ill but using your awesome work to assign it's mac to the switch instead.

Thanks again

**Reply by grimpy** [reply]
Meanwhile I've cleaned up the code in the gist a little bit with the help of @ndeadly, I just wanted to add a final note if you want to use this on a linux system you need to reverse the retrieved key for instance: AB CD EF would become EF CD AB.

**Reply by ndeadly** [reply]
Looks much cleaner. Were you able to successfully sync the pairings?

**Reply by grimpy** [reply]
Yes, 

Since my PC refused to take a new mac I also used mission control to spoof the mac on the switch itself.
Then on the PC I edited `/var/lib/bluetooth/<btadaptermac>/<gamepad mac>/info` restart bluetooth daemon and all was well,

**Reply by ndeadly** [reply]
Nice.
